### ISSUE TITLE
"Apply" very slow

### ISSUE THREAD

**Issue by NicolaP8** [ukanth/afwall#793]
Hi,
we don't need to see every number when applying, it's enough an update every half second!
This will speed up a lot :)
Now with 304 rules I've to wait 13 seconds (Moto G5 Plus).
Thanks 
         N

**Comment by coolicer** [comment]
Can it make a cache file,only change the new one?

**Comment by ildar** [comment]
isn't the issue solved with latest release?

**Comment by ukanth** [comment]
No, this issue is not yet resolved. The main problem is iptables itself synchronous (you can execute only one at a time).

**Comment by skweezyweezulz** [comment]
"The main problem is iptables itself [is] synchronous (you can execute only one at a time)."

While this is true of the current mechanism that inserts rules one at a time via the standard iptables command line interface, this is not true when you use the iptables-restore and iptables-save interfaces as well as their equivalents for IPv6.

A quick look at iptables-restore.c and ip6tables-restore.c for iptables version 1.4.20, which is what's currently stored in your GitHub tree, shows that the input read from stdin is actually built up in memory and applied as a single commit operation. This mechanism doesn't require a separate module .ko file to insmod and in fact depends on only the standard ip_tables and ip6_tables kernel modules which should be present for any Android version that supports a firewall.

So while what you say is true, there isn't really an excuse for not rearchitecting the solution in a way where all of the rules could be applied through a single commit operation.

I'll add my own abysmal time while we're on this subject: approximately 2650 rules took roughly 145 seconds to apply, with no exclusive locking during this time, meaning that if you interrupt this process, chances are very good you will have duplicate rules on the next apply.

This is awful enough to make this software unusable for complex configurations.

Provided that iptables-restore and ip6tables-restore will function properly on a wide range of Android versions, and given the code I've seen I see no reason that this wouldn't be the case, this should be not only the preferred way to implement a speed-up but also the preferred way to update the firewall at any time. It's considerably less prone to errors than inserting rules with an iterator and hoping for the best.

I realise developer time is as always a finite resource, but this really would be the best way to improve this project.

**Comment by ukanth** [comment]
@skweezyweezulz , I did look at iptables-restore and other mechanisms too. As I explained earlier, I could put all the iptables rules in script and execute at once or using iptables-restore ( it has it's own limitation when comes to android related to uid handling).

But as per the current design iptables rules are executed one by one and if it fails due to external factors like network change and then it will try applying the same rule again instead of simply fail or skipping. That makes things slower. So ultimately it's stability vs speed at the end.
